### PR TITLE
Explicitly state that Starlark syntax is strict subset of Python

### PR DESCRIPTION
**Pull request by stepancheg**
State that Starlark syntax (but not semantic) is a strict subset of Python, and this property will be maintained in the future iterations of Starlark.

Context: users have asked what tools they can use to work with starlark sources to implement codemods, linters, code formatters, syntax highlighters, documentation generators etc.

One possible option is to recommend Python tools when AST is needed, for example, Python's builtin `ast.parse`. However, to make that recommendation correct, we should guarantee that Starlark syntax (but not semantics) will stay strict subset of Python: so each valid Starlark program could be parsed as Python, and tools won't need to be heaviy rewritten when Starlark language changes.

This PR automatically closes
https://github.com/bazelbuild/starlark/issues/111
as it contravenes this newly established requirement.

### CODE REVIEW

**Comment by laurentlb** [comment]
> One possible option is to recommend Python tools when AST is needed

We have multiple Starlark parsers and I would recommend using them instead of Python, when possible.

For formatting, I strongly recommend Buildifier as I would prefer to have only one standard formatting.

For AST manipulation, the recommendation is also to rely on https://github.com/bazelbuild/buildtools

**Comment by stepancheg** [comment]
> For formatting, I strongly recommend Buildifier as I would prefer to have only one standard formatting.

There are several issues with buildifier:
* it does not support starlark-rust extensions
* how it formats code, and depends on your preferences, it might seem not as good as Black

> For AST manipulation, the recommendation is also to rely on https://github.com/bazelbuild/buildtools

It is written in Go, so if it is not your default choice for tools, it might be slightly inconvenient or very inconvenient (for example, if you want to use some API in addition to working with AST, and that API provided in Python).

Also I suspect that LibCST (which I used recently to do codemods) seems to do better job of preserving whitespaces, so it is better suitable to do codemods on files unless files were already formatted with buildifier.

Or for example, if you do codemod, you can ask ChatGPT to generate the full program for the codemod for you with libcst, but not for buildifier.

I'm saying this to illustrate that different tools might be needed depending on use case.

Anyway, my point is, being able to pick from variety of tools is better that sticking to one repository. By restricting syntax to Python, starlark gains a lot and loses very little.

**Comment by laurentlb** [comment]
I agree with "restricting syntax to Python" and I also think it's very valuable.

Using LibCST seems reasonable to me. The AST in the Buildifier library is deliberately simpler (without the spaces) because our approach is to require all Starlark files to be formatter with Buildifier. For this reason, I'd rather not mention Black here.

It could be useful to discuss (in a separate thread) what Starlark-Rust extensions could be standardized.

To sum up, I think it's a good PR, but we can remove Black from the text. I'll wait a bit and see if anyone else wants to chime in.

**Comment by stepancheg** [comment]
Removed references to tool names, replaced with generic phrase "Practically it means, tools working with Python AST can be used to work with Starlark files."

**Comment by stepancheg** [comment]
Nobody objected in a week. Let's land it?

**Comment by laurentlb** [comment]
Agreed.
I cannot merge now though, because a test is failing (I haven't looked at it)

**Comment by stepancheg** [comment]
https://github.com/bazelbuild/starlark/pull/259

**Review by laurentlb — APPROVED** [review]
